src/nvim/popupmenu.c

```diff
@@ -660,6 +662,131 @@ void pum_redraw(void)
   }
 }
 
+static win_T *pum_create_float_preview(void)
+{
+  FloatConfig config = FLOAT_CONFIG_INIT;
+  config.relative = kFloatRelativeEditor;
+  // when pum_above is SW otherwise is NW
+  config.anchor = pum_above ? kFloatAnchorSouth : 0;
+  int col = pum_col + pum_width + pum_scrollbar + 1;
+  // TODO(glepnir): support config align menu or item
+  // align menu
+  config.row = pum_row;
+  int right_remain = Columns - col;
+
+  if (right_remain >= 20) {
+    config.width = right_remain;
+    config.col = col - 1;
+  } else if (right_remain < 20 && pum_col >= 20) {
+    config.width = pum_col - 2;
+    config.col = pum_col - config.width - 1;
+  } else {
+    return NULL;
+  }
+  config.height = pum_height;
+  config.style = kWinStyleMinimal;
+  config.hide = true;
+  Error err = ERROR_INIT;
+  win_T *wp = win_new_float(NULL, true, config, &err);
+  win_enter(wp, false);
```

Comment:
do we need to enter window? Just displaying some text and setting a few basic options should be possible without entering.

Reply:
in `pum_set_selected` need `win_enter` . in api level maybe no need enter. I can add a bool flag .

Reply:
try use `wp->w_buffer` and remove `win_enter` in there will cause the curbuf changed wired.

Reply:
done.